projects/storefrontlib/src/shared/components/popover/popover.directive.ts

```diff
@@ -0,0 +1,240 @@
+      this.open(event);
+    else if (this.isOpen) this.close();
+  }
+
+  /**
+   * Method creates instance and pass parameters to popover component.
+   */
+  renderPopover() {
+    const containerFactory = this.componentFactoryResolver.resolveComponentFactory(
+      PopoverComponent
+    );
+    this.popoverContainer = this.viewContainer.createComponent(
+      containerFactory
+    );
+    if (this.popoverContainer && this.popoverContainer.instance) {
+      this.popoverContainer.instance.content = this.cxPopover;
+      this.popoverContainer.instance.triggerElement = this.element;
+      this.popoverContainer.instance.popoverInstance = this.popoverContainer;
+      this.popoverContainer.instance.position = this.placement;
+      this.popoverContainer.instance.customClass = this.customClass;
+      this.popoverContainer.instance.focusConfig = this.focusConfig;
+      this.popoverContainer.instance.appendToBody = this.appendToBody;
+      this.popoverContainer.instance.positionOnScroll = this.positionOnScroll;
+
+      if (this.appendToBody) {
+        this.renderer.appendChild(
+          this.document.body,
+          this.popoverContainer.location.nativeElement
+        );
+      }
+
+      this.popoverContainer.changeDetectorRef.detectChanges();
+    }
+  }
+
+  /**
+   * Method uses `Renderer2` service to listen every click event.
+   *
+   * Registered only when popover state was set to open and checks if element
+   * outside popover component was clicked.
+   *
+   * If so directive performs `close()` action and fire unlistener for such event.
+   */
+  triggerClickEvent() {
+    this.clickEventUnlistener = this.renderer.listen(
+      this.winRef.nativeWindow,
+      'click',
+      (event: Event) => {
+        if (event && event.target !== this.element.nativeElement) this.close();
+      }
+    );
+  }
+
+  /**
+   * Method uses `Renderer2` service to listen every escape keydown event.
+   *
+   * Registered only when popover state was set to open and checks if `escape`
+   * key was pressed.
+   *
+   * If so directive performs `close()` action and fire unlistener for such event.
+   */
+  triggerEscKeydownEvent() {
+    this.keydownEventUnlistener = this.renderer.listen(
+      this.winRef.nativeWindow,
+      'keydown.escape',
+      (event: KeyboardEvent) => {
+        if (event) {
+          this.close();
+          if (this.focusConfig && this.appendToBody) {
+            this.element.nativeElement.focus();
+          }
+        }
+      }
+    );
+  }
+
+  constructor(
+    protected element: ElementRef,
+    protected viewContainer: ViewContainerRef,
+    protected componentFactoryResolver: ComponentFactoryResolver,
+    protected renderer: Renderer2,
+    protected changeDetectorRef: ChangeDetectorRef,
+    protected positioningService: PositioningService,
+    protected popoverService: PopoverService,
+    protected winRef: WindowRef,
+    @Inject(DOCUMENT) protected document: any
```

Comment:
winRef already has Document token injected

Reply:
ok, removed that from constructor